logstash-core/lib/logstash/plugins/registry.rb

```diff
@@ -1,86 +1,243 @@
+          end
         end
-        return klass
-      else
-        # The plugin was defined directly in the code, so there is no need to use the
-        # require way of loading classes
-        return @registry[plugin.path] if registered?(plugin.path)
-        raise LoadError
       end
-    rescue => e
-      @logger.warn("Problems loading a plugin with", :type => type, :name => plugin, :path => plugin.path,
-                   :error_message => e.message, :error_class => e.class, :error_backtrace => e.backtrace)
-      raise LoadError, "Problems loading the requested plugin named #{plugin_name} of type #{type}. Error: #{e.class} #{e.message}"
     end
 
-    def register(path, klass)
-      @registry[path] = klass
+    def lookup(type, plugin_name, &block)
+      plugin = get(type, plugin_name)
+      # Assume that we have a legacy plugin
+      if plugin.nil?
+        begin
+          path = "logstash/#{type}s/#{plugin_name}"
+
+          begin
+            require path
+          rescue LoadError
+            # Plugin might be already defined in the current scope
+            # This scenario often happen in test when we write an adhoc class
+          end
+
+          klass = namespace_lookup(type, plugin_name)
+          plugin = lazy_add(type, plugin_name, klass)
+        rescue => e
+          logger.warn("Problems loading a plugin with",
+                      :type => type,
+                      :name => plugin_name,
+                      :path => path,
+                      :error_message => e.message,
+                      :error_class => e.class,
+                      :error_backtrace => e.backtrace)
+
+          raise LoadError, "Problems loading the requested plugin named #{plugin_name} of type #{type}. Error: #{e.class} #{e.message}"
+        end
+      end
+
+      if block_given? # if provided pass a block to do validation
+        raise LoadError unless block.call(plugin.klass, plugin_name)
+      end
+
+      return plugin.klass
+    end
+
+    def lookup_pipeline_plugin(type, name)
+      LogStash::PluginRegistry.lookup(type, name) do |plugin_klass, plugin_name|
+        is_a_plugin?(plugin_klass, plugin_name)
+      end
+    rescue LoadError, NameError => e
+      logger.debug("Problems loading the plugin with", :type => type, :name => name)
+      raise(LogStash::PluginLoadingError, I18n.t("logstash.pipeline.plugin-loading-error", :type => type, :name => name, :error => e.to_s))
+    end
+
+    def lazy_add(type, name, klass)
+      logger.error("On demand adding plugin to the registry", :name => name, :type => type, :class => klass)
+      add_plugin(type, name, klass)
+    end
+
+    def add(type, name, klass)
+      logger.error("Adding plugin to the registry", :name => name, :type => type, :class => klass)
+      add_plugin(type, name, klass)
```

Comment:
Yes, I think for debugging purpuse and since we support both legacy declaration of plugins and this new hook style registration It make sense to know when this plugin get added to the registry.